```
  1) workers.social.group.index #destroy() create a team should add resources, invitations, apiToken, groupData to team and add slack integration to admin account:
     Uncaught AssertionError: expected null to exist
      at workers/social/lib/social/models/group/index.test.coffee:303:29
      at workers/social/lib/social/models/group/groupdata.coffee:104:9
      at readableAddChunk (_stream_readable.js:176:18)
      at Socket.Readable.push (_stream_readable.js:134:10)
      at TCP.onread (net.js:548:20)
```

--- a/servers/lib/server/helpers/oauth.coffee
+++ b/servers/lib/server/helpers/oauth.coffee
@@ -39,6 +39,7 @@ fetchGroupOAuthSettings = (provider, clientId, state, callback) ->
           redirectUri: "#{protocol}//#{slug}.#{hostname}/-/oauth/#{provider}/callback"
         }
 
+
 saveOauthToSession = (oauthInfo, clientId, provider, callback) ->
   { JSession } = koding.models
 
@@ -51,6 +52,7 @@ saveOauthToSession = (oauthInfo, clientId, provider, callback) ->
   query["foreignAuth.#{provider}"] = oauthInfo
 
   JSession.update { clientId }, { $set:query }, callback
+
 
 # Get user info with access token
 fetchUserOAuthInfo = (provider, req, res, data) ->

--- a/website/swagger.json
+++ b/website/swagger.json
@@ -8023,7 +8023,7 @@
         "consumes": [
           "application/json"
         ],
-        "description": "",
+        "description": "Method JGroup.fetchDataAt",
         "parameters": [
           {
             "$ref": "#/parameters/instanceParam"

--- a/workers/social/lib/social/models/group/groupdata.coffee
+++ b/workers/social/lib/social/models/group/groupdata.coffee
@@ -23,9 +23,8 @@ module.exports = class JGroupData extends Model
 
     sharedMethods :
       static      :
-        fetchByKey: [
-          (signature String, Function)
-        ]
+        fetchByKey: (signature String, Function)
+
 
   @create = (slug, callback) ->
     data = new JGroupData { slug }
@@ -38,8 +37,9 @@ module.exports = class JGroupData extends Model
       return callback err  if err
       callback null, data
 
-  @fetchData: (slug, callback) ->
-    return callback new Error 'slug is required' unless slug
+
+  @fetchData = (slug, callback) ->
+    return callback new Error 'slug is required'  unless slug
 
     JGroupData.one { slug }, (err, data) ->
       return callback err  if err
@@ -65,8 +65,8 @@ module.exports = class JGroupData extends Model
     JGroupData.fetchDataAt slug, path, callback
 
 
-  @fetchDataAt: (slug, path, callback) ->
-    return callback new Error 'slug is required' unless slug
+  @fetchDataAt = (slug, path, callback) ->
+    return callback new Error 'slug is required'  unless slug
 
     opts = {}
     opts["payload.#{path}"] = 1 if path
@@ -77,22 +77,25 @@ module.exports = class JGroupData extends Model
       payload = if data then getAt data.payload, path else null
       return callback null, payload
 
+
   # see docs on JGroup::modifyData
-  @modifyData : (slug, data, callback) ->
-    return callback new Error 'slug is required' unless slug
+  @modifyData = (slug, data, callback) ->
+    return callback new Error 'slug is required'  unless slug
 
     # it's only allowed to change followings
-    whitelist  = ['github.organizationToken', 'test_key__' ]
+    allowedPaths = [ 'github.organizationToken', 'test_key__' ]
 
     # handle $set and $unset cases in one
-    operation  = { $set: {}, $unset: {} }
+    operation = {}
 
-    for item in whitelist
-      key = "payload.#{item}"
-      val = getAt data, item
-      if val
-      then operation.$set[key]   = val
-      else operation.$unset[key] = ''
+    allowedPaths.forEach (path) ->
+      key = "payload.#{path}"
+      if val = data[path]
+        operation.$set ?= {}
+        operation.$set[key] = val
+      else
+        operation.$unset ?= {}
+        operation.$unset[key] = ''
 
     JGroupData.fetchData slug, (err, data) ->
       return callback err  if err

--- a/workers/social/lib/social/models/group/index.coffee
+++ b/workers/social/lib/social/models/group/index.coffee
@@ -554,15 +554,17 @@ module.exports = class JGroup extends Module
     JGroupData = require './groupdata'
     JGroupData.fetchData @slug, callback
 
+  fetchDataAt: (path, callback) ->
+    JGroupData = require './groupdata'
+    JGroupData.fetchDataAt @slug, path, callback
+
   fetchDataAt$: permit
     advanced: [
       { permission: 'grant permissions' }
       { permission: 'grant permissions', superadmin: yes }
     ]
     success: (client, path, callback) ->
-      JGroupData = require './groupdata'
-      JGroupData.fetchDataAt @slug, path, callback
-
+      @fetchDataAt path, callback
 
   sendNotification: (event, contents, callback) ->
     JGroup.sendNotification @getAt('slug'), event, contents, callback
@@ -1048,10 +1050,9 @@ module.exports = class JGroup extends Module
       { permission: 'edit own groups', validateWith : Validators.group.admin }
       { permission: 'edit groups',     superadmin   : yes }
     ]
-    success  : (client, _data, callback) ->
+    success  : (client, data, callback) ->
       JGroupData = require './groupdata'
-      JGroupData.modifyData @slug, _data, (err) ->
-        return callback err
+      JGroupData.modifyData @getAt('slug'), data, callback
 
 
   modify     : permit
@@ -1127,7 +1128,7 @@ module.exports = class JGroup extends Module
         return callback err  if err
 
         dataToUnset = {}
-        dataToUnset["data.#{provider}.applicationSecret"] = null
+        dataToUnset["payload.#{provider}.applicationSecret"] = null
         data.update { $unset: dataToUnset }, callback
 
 
@@ -1181,7 +1182,7 @@ module.exports = class JGroup extends Module
             return callback err  if err
 
             dataToSet = {}
-            dataToSet["data.#{provider}"] = { applicationSecret }
+            dataToSet["payload.#{provider}"] = { applicationSecret }
             data.update { $set: dataToSet }, (err) ->
               return callback err  if err
 

--- a/workers/social/lib/social/models/group/index.test.coffee
+++ b/workers/social/lib/social/models/group/index.test.coffee
@@ -300,7 +300,7 @@ runTests = -> describe 'workers.social.group.index', ->
 
           (next) ->
             group.modifyData client, { 'foo': 'bar' }, (err) ->
-              expect(err).to.exist
+              expect(err).to.not.exist
               next()
 
           (next) ->

